tracee/tracee.bpf.c

```diff
@@ -497,13 +500,35 @@ static __always_inline int get_config(u32 key)
     return *config;
 }
 
+// returns 1 if you should trace based on uid, 0 if not
 static __always_inline int uid_filter_matches()
 {
     uid_t uid = (u32)bpf_get_current_uid_gid();
-    u32* isMatch = bpf_map_lookup_elem(&uid_filter, &uid);
-    if (isMatch == NULL) {
+    s64* equality = bpf_map_lookup_elem(&uid_equal_filter, &uid);
+    if (equality != NULL) {
+        return *equality;
+    }
+
+    u32 greater = GREATER;
+    u32 less = LESS;
+
+    s64* greaterThan = bpf_map_lookup_elem(&uid_compare_filter, &greater);
+    s64* lessThan = bpf_map_lookup_elem(&uid_compare_filter, &less);
+
+    if (lessThan != NULL) {
+        if (uid > *lessThan)
+            return 0;
+    }
+
+    if (greaterThan != NULL) {
+        if (uid < *greaterThan)
+            return 0;
+    }
+
+    if (greaterThan == NULL && lessThan == NULL) {
```

Comment:
Ok, so I reproduced the verifier issue with kernel 5.4.
For some reason, it doesn't like this line.
Then I thought - maybe we can remove this last if.

But then we come back to the problem you said:
if uid=5 was given as a filter, and now we have an event with uid=7, then it should not pass the filter.
The problem is, that in this case, we want the defaults to be "close" for inequality.

So what I suggest is the following:
1. Remove this if and pass the verifier
2. If ONLY equality filter was given (no greater than, no less than, no !=) - default to close values
3. Otherwise, default to open values.

This is logic, I think, because it should satisfy the equation you wrote:
 (=) || ((<) && (>) && !(!=))

WDYT?

Reply:
Glad you were able to reproduce. Yes that logic makes sense to me, working on implementing that and taking into account all your previous comments. I'm surprised this problem become so complex to be honest.

Reply:
I'm also surprised about that. But I think two things made it complex:
1. A comment I wrote few days ago:
"This is more intuitive. It just makes it a bit harder to implement - mixing AND and OR together..."
2. When writing bpf code, you can never know where the verifier will block you (unless you are one of the developers of the verifier in the kernel)

Reply:
Codifying user intent always can be tough. There's a lot of work to get done for CI/CD, testing against various kernel/verifier versions. That should simplify catching future issues.